### PR TITLE
simplify processing of ConstVal objects when not all variants are legal

### PR DESCRIPTION
**Pull request by oli-obk**
r? @eddyb 

Adding new variants is annoying as one needs to modify all these places that **don't** handle the new variant.

I chose not to use `Display` as I don't think it is appropriate.

### CODE REVIEW

**Comment by eddyb** [comment]
Why not add another `enum` instead of using strings?

**Comment by oli-obk** [comment]
No reason in particular...

I'd love to have a repr C discriminant enum for every enum and an easy way to convert to it... (oh + subset-enums xD ) but since we don't have that, I'll just hand-code them

**Comment by oli-obk** [comment]
hmm actually there is a reason... another enum will require touching all those places again and only save us from modifying the hand-flattened `ErrorKind` enum... What I could do is dump the actual `ConstVal` into the type instead of a string

File: src/librustc/middle/ty.rs [hunk 1]
```diff
@@ -6100,13 +6100,7 @@ impl<'tcx> ctxt<'tcx> {
...
                     ConstVal::Uint(count) => return count as usize,
                     ConstVal::Int(count) if count >= 0 => return count as usize,
-                    ConstVal::Int(_) => "negative integer",
-                    ConstVal::Float(_) => "float",
-                    ConstVal::Str(_) => "string",
-                    ConstVal::Bool(_) => "boolean",
-                    ConstVal::Binary(_) => "binary array",
-                    ConstVal::Struct(..) => "struct",
-                    ConstVal::Tuple(_) => "tuple"
+                    const_val => const_val.description(),
```
Comment (eefriedman) [inline src/librustc/middle/ty.rs]: You might want to keep around the special-case for ConstVal::Int?  The fact that the integer is negative is important.

Comment (oli-obk) [inline src/librustc/middle/ty.rs]: i'll just change the description method to emit that information, too

**Comment by oli-obk** [comment]
updated to pass around the actual `ConstVal` object instead of a string and to print "negative integer" for negative signed integers

**Comment by oli-obk** [comment]
ping @eddyb

**Pull request merged**